OSDOCS-14389:CUDN best practices localnet (#93383)
openshift/openshift-docs: modules/nw-cudn-best-practices.adoc

** If the namespace is missing the `k8s.ovn.org/primary-user-defined-network` label and a primary `ClusterUserDefinedNetwork` CR already exists, a pod in the namespace is created and attached to the default network.

> ** If the namespace _has_ the label, and a primary `ClusterUserDefinedNetwork` CR does not exist, a pod in the namespace is not created until the `ClusterUserDefinedNetwork` CR is created.

Reviewer: It may sound better to replace "a pod in the namespace is not created until (...)" with "any pod creation in the namespace will remain stuck until (...)"

* When using the `ClusterUserDefinedNetwork` CR to create `localnet` topology, the following are best practices for administrators:

> ** You must make sure that the `spec.network.physicalNetworkName` field matches the field that you configured in the OVS bridge mapping when you create your CUDN. This ensures that you are bridging to the intended segment of your physical network. Do not reuse the same `physicalNetworkName` for different physical networks.

Reviewer 2: We actually do support multiple CUD  CRs using the same bridge-mapping.
@maiqueb right?

> ** You must make sure that the `spec.network.physicalNetworkName` field matches the field that you configured in the OVS bridge mapping when you create your CUDN. This ensures that you are bridging to the intended segment of your physical network. Do not reuse the same `physicalNetworkName` for different physical networks.

Reviewer 3: I've tried to say that in https://github.com/openshift/openshift-docs/pull/93383#discussion_r2092868358

> ** You must make sure that the `spec.network.physicalNetworkName` field matches the field that you configured in the OVS bridge mapping when you create your CUDN. This ensures that you are bridging to the intended segment of your physical network. Do not reuse the same `physicalNetworkName` for different physical networks.

Author: I reworded to include phrasing that indicates that we support multiple CUDN CRs using the same bridge mapping.

> ** You must make sure that the `spec.network.physicalNetworkName` field matches the one configured in the OVS bridge mapping when creating your CUDN. This ensures that you are bridging to the segment of your physical network that you intended to bridge to. Avoid reusing the same `physicalNetworkName` for different physical networks.

Reviewer 3: I don't think "avoid" is the proper word to be used. 

In my understanding, the admin must be aware they are essentially *defining* how the network is when they use they attribute. I.e. if they re-use the `physicalNetworkName`, they are essentially **connecting** to that network. I.e. they are not "different physical networks" .

> ** You must make sure that the `spec.network.physicalNetworkName` field matches the one configured in the OVS bridge mapping when creating your CUDN. This ensures that you are bridging to the segment of your physical network that you intended to bridge to. Avoid reusing the same `physicalNetworkName` for different physical networks.

Reviewer 4: Minor suggestions to avoid the use of gerunds and for concision.

```suggestion
** You must make sure that the `spec.network.physicalNetworkName` field matches the field that you configured in the OVS bridge mapping when you create your CUDN. This ensures that you are bridging to the intended segment of your physical network. Do not reuse the same `physicalNetworkName` for different physical networks.
```

> ** Avoid overlapping subnets between your physical network and your other network interfaces. Overlapping network subnets can cause routing conflicts and network instability.

Reviewer 2: I think it should mention that when OVN-K IPAM is being used (i.e.: when `spec.nework.localnet.subsets` is specified) one could prevent such conflicts by specifying the reserved IPs/subnets in the `spec.nework.localnet.excludeSubnets` attribute.

Author: I rewrote to try to clarify.

Author: @maiqueb here is comment where I introduced the mistake came from. I'll fix, sorry.

> ** Avoid overlapping subnets between your physical network and your other network interfaces. Overlapping network subnets can cause routing conflicts and network instability. To prevent conflicts when using the `spec.network.localnet.joinSubnets` parameter, you might use the `spec.network.localnet.excludeSubnets` parameter.

Author: Sorry, this is typo and confusion between the last round of comments. I think this should just be subnets.

> ** When configuring a Virtual Local Area Network (VLAN), you must ensure that both your underlying physical infrastructure (switches,routers) as well as your nodes are properly configured to accept VLAN IDs (VIDs). This means configuring the physical network interface (e.g. `eth1`) as an access port for the VLAN (e.g. 20) you are connecting to through the physical switch. In addition, you must verify that an OVS bridge mapping (e.g. `eth1`) exists on your nodes to ensure that that the physical interface is properly connected with OVN-Kubernetes.

Reviewer 3: wait, why isn't using a trunk on the physical interface an option ? I think it is a valid use-case - but the physical underlay (i.e. the physical switch config) must conform to this configuration (i.e. the VLAN IDs used in the trunk **must** be defined in the switch .

Author: I don't think I fully understand the `trunk` value so I'm not sure I can fully capture the use case. You raise a good point though and I wonder if we should update the CUDN CR example or at least the `Additional config details` for CUDN to include this information.

WDYT?

> ** When configuring a Virtual Local Area Network (VLAN), you must ensure that both your underlying physical infrastructure (switches,routers) as well as your nodes are properly configured to accept VLAN IDs (VIDs). This means configuring the physical network interface (e.g. `eth1`) as an access port for the VLAN (e.g. 20) you are connecting to through the physical switch. In addition, you must verify that an OVS bridge mapping (e.g. `eth1`) exists on your nodes to ensure that that the physical interface is properly connected with OVN-Kubernetes.

Reviewer 4: Minor suggestions to avoid the use of gerunds and for concision (the[ ISG](https://www.ibm.com/docs/en/ibm-style?topic=word-usage) suggests 'and' instead of 'as well as'). Also, the [ ISG](https://www.ibm.com/docs/en/ibm-style?topic=word-usage) suggests 'for example' instead of 'e.g.'.

```suggestion
** When you configure a Virtual Local Area Network (VLAN), you must ensure that both your underlying physical infrastructure (switches, routers, and so on) and your nodes are properly configured to accept VLAN IDs (VIDs). This means that you configure the physical network interface, for example `eth1`, as an access port for the VLAN, for example 20, that you are connecting to through the physical switch. In addition, you must verify that an OVS bridge mapping, for example `eth1`, exists on your nodes to ensure that that the physical interface is properly connected with OVN-Kubernetes.
```

> ** When you configure a Virtual Local Area Network (VLAN), you must ensure that both your underlying physical infrastructure (switches, routers, and so on) and your nodes are properly configured to accept VLAN IDs (VIDs). This means that you configure the physical network interface, for example `eth1`, as an access port for the VLAN, for example `20`, that you are connecting to through the physical switch. In addition, you must verify that an OVS bridge mapping, for example `eth1`, exists on your nodes to ensure that that the physical interface is properly connected with OVN-Kubernetes.

Reviewer: Nobody expands the VLAN abbreviation. If we must do it due to stylistic guidelines, then OK, but for many readers it might sound a bit pedantic.
